Compiler incorrectly (and inconsistently) finds Iterator methods on i32

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur. 
-->

Given the following code: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=cb84fd14f11aca6ad741f9537a75c241)

```rust
fn main() {
    let x: i32 = 1;
    println!("{:?}", x.count())
}
```

The current output is:

```
error[E0599]: the method `count` exists for type `i32`, but its trait bounds were not satisfied
 --> src/main.rs:3:24
  |
3 |     println!("{:?}", x.count())
  |                        ^^^^^ method cannot be called on `i32` due to unsatisfied trait bounds
  |
  = note: the following trait bounds were not satisfied:
          `i32: Iterator`
          which is required by `&mut i32: Iterator`

error: aborting due to previous error
```

<!-- The following is not always necessary. -->
Ideally the output should look like:

```
error[E0599]: the method `count` does not exist for type `i32`
```

For some reason, this doesn't happen for all traits in `std`. Eg

```rust
fn main() {
    let x: i32 = 1;
    println!("{:?}", x.write(&[1, 2, 3]))
}
```

Just gives 
```
error[E0599]: no method named `write` found for type `i32` in the current scope
 --> src/main.rs:3:24
  |
3 |     println!("{:?}", x.write(&[1, 2, 3]))
  |                        ^^^^^ method not found in `i32`

error: aborting due to previous error
```
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=57a3dc905035568e0161bad29f29582f)


Version: 1.53.0-nightly (2021-04-22 7f4afdf0255600306bf6)
<!--
If the problem is not self-explanatory, please provide a rationale for the
change.
-->

<!--
If dramatically different output is caused by small changes, consider also
adding them here.

If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions. The output might also be
different depending on the Edition.
-->

Some discussion on Discord suggests that
```rust
impl<I: Iterator + ?Sized> Iterator for &mut I
```
[src](https://doc.rust-lang.org/src/core/iter/traits/iterator.rs.html#3392-3406)

might be a culprit.

Trying with the variant
```rust
fn main() {
    let x: i32 = 1;
    println!("{:?}", x.next())
}
```
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1af44d169ae295ed19cdfde82d75a067)

produces
```
error[E0599]: no method named `next` found for type `i32` in the current scope
 --> src/main.rs:3:24
  |
3 |     println!("{:?}", x.next())
  |                        ^^^^ method not found in `i32`

error: aborting due to previous error
```
as expected.

So it looks like the trait bound `I: Iterator + ?Sized` is being honored for the search for the `next` method, which is defined in the `impl<…> Iterator for &mut I` block, but not for the search for the `count` method, which is defined in `trait Iterator` definition.

I tried to replicate the incorrect diagnostic with a more minimal case than `Iterator`. It looks like it depends on the borrow mode of the receiver. This code [(playground)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=04e0f256725c95478bad60ab971d67fd):

```rust
trait Trait {
    // similar to Iterator::next
    fn f1(&self);
    // similar to Iterator::count
    fn f2(self) where Self: Sized {}
    // similar to Iterator::size_hint
    fn f3(&self) {}
}
// similar to impl<T: Iterator + ?Sized> Iterator for &mut T
impl<T: Trait> Trait for &T
{
    fn f1(&self) { (**self).f1() }
}
fn main() {
    1i32.f1();
    1i32.f2();
    1i32.f3();
}
```

produces these errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0599]: no method named `f1` found for type `i32` in the current scope
  --> src/lib.rs:15:10
   |
15 |     1i32.f1();
   |          ^^ method not found in `i32`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
note: `Trait` defines an item `f1`, perhaps you need to implement it
  --> src/lib.rs:1:1
   |
1  | trait Trait {
   | ^^^^^^^^^^^

error[E0599]: the method `f2` exists for type `i32`, but its trait bounds were not satisfied
  --> src/lib.rs:16:10
   |
16 |     1i32.f2();
   |          ^^ method cannot be called on `i32` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `i32: Trait`
           which is required by `&i32: Trait`
   = help: items from traits can only be used if the trait is implemented and in scope
note: `Trait` defines an item `f2`, perhaps you need to implement it
  --> src/lib.rs:1:1
   |
1  | trait Trait {
   | ^^^^^^^^^^^

error[E0599]: no method named `f3` found for type `i32` in the current scope
  --> src/lib.rs:17:10
   |
17 |     1i32.f3();
   |          ^^ method not found in `i32`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
note: `Trait` defines an item `f3`, perhaps you need to implement it
  --> src/lib.rs:1:1
   |
1  | trait Trait {
   | ^^^^^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

`Trait::f1()` is a required trait method that takes a ref to `self`, similar to `Iterator::next()` (the `mut` doesn't seem necessary to exhibit these diagnostics).
`Trait::f2()` is a default trait method that moves `self`, similar to most of the `Iterator` methods. It seems like a default trait method that moves `self` is necessary to trigger this incorrect diagnostic.
`Trait::f3()` is a default trait method that takes a ref to `self`, similar to `Iterator::size_hint()`.

I've built a debugging version of the compiler and I'm trying to understand the debug logs for the trait and typechecking systems.

--- a/src/test/ui/methods/issues/issue-84495.rs
+++ b/src/test/ui/methods/issues/issue-84495.rs
@@ -1,0 +1,4 @@
+fn main() {
+    let x: i32 = 1;
+    println!("{:?}", x.count()); //~ ERROR is not an iterator
+}

--- a/src/test/ui/methods/issues/issue-84495.stderr
+++ b/src/test/ui/methods/issues/issue-84495.stderr
@@ -1,0 +1,13 @@
+error[E0599]: `i32` is not an iterator
+  --> $DIR/issue-84495.rs:3:24
+   |
+LL |     println!("{:?}", x.count());
+   |                        ^^^^^ `i32` is not an iterator
+   |
+   = note: the following trait bounds were not satisfied:
+           `i32: Iterator`
+           which is required by `&mut i32: Iterator`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0599`.

